### PR TITLE
feat: support absolute paths when matching paths

### LINKED ISSUE #7350: `forge test` doesn't support absolute path
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (7922fd5 2024-02-05T00:19:20.258041000Z)

### What command(s) is the bug in?

forge test

### Operating System

macOS (Intel)

### Describe the bug

`forge test --match-path  absolute_path` will report `No tests match the provided pattern`, only relative path works.

` --match-path` should work for both absolute path and relative path.

### PR DESCRIPTION
**Pull request by matthewliu10**
## Motivation

Closes #7350

forge test --match-path  absolute_path will report No tests match the provided pattern, only relative path works.

 --match-path should work for both absolute path and relative path.

## Solution

If the path given by the user is absolute, and the path given to the function is relative, it uses the canonicalize() function to turn the path given to the function into an absolute path so that the paths can be compared.

## Notes

I initially tried converting both paths into their canonical forms using canonicalize and directly comparing them, however this failed on the can_test_with_match_path test because it would attempt to canonicalize "*src/ATest.t.sol" which causes an error.

There may be cases where the paths refer to the same location but is_match() still returns false. For example, if the user inputs "/home/user/a/b/../..//c/", it may not match with "/home/user/c/" even though they are equivalent. Note: canonicalize() resolves any symbolic links, relative path components, and removes any redundant components in the path.

## Tests

The following 5 tests fail when running cargo test --all --all-features:

cmd::can_install_missing_deps_build
cmd::can_install_missing_deps_test
cmd::can_update_library_with_outdated_nested_dependency
config::can_parse_default_fs_permissions
ext_integration::snekmate

However, the following 4 tests still failed when I reverted the changes that I made (reverted code to the original code that I pulled from the repo):

cmd::can_install_missing_deps_build
cmd::can_install_missing_deps_test
cmd::can_update_library_with_outdated_nested_dependency
ext_integration::snekmate

The following test failed when I ran cargo test --all --all-features, but passed when I ran it individually:

config::can_parse_default_fs_permissions

### CODE REVIEW

**Comment by matthewliu10** [comment]
The test that is failing is the new one that I added. When testing locally, it is passing:
image

File: crates/common/src/glob.rs [hunk 1]
```diff
@@ -46,6 +46,16 @@ impl GlobMatcher {
         if !path.starts_with("./") && self.as_str().starts_with("./") {
             return self.matcher.is_match(format!("./{}", path.display()));
         }
+        let glob_path = PathBuf::from(self.glob().glob());
+        if path.is_relative() && glob_path.is_absolute() {
+            let canonicalized_path = match path.canonicalize() {
```
Comment (klkvr) [inline crates/common/src/glob.rs:51]:
```suggestion
            let canonicalized_path = match dunce::canonicalize(path) {
```

Comment (klkvr) [reply crates/common/src/glob.rs:51]: The test is failing only on Windows, this should probably fix it.

**Review by klkvr — CHANGES_REQUESTED** [review]

**Comment by matthewliu10** [comment]
I've just run the failing tests locally. Every test either fails when using the code from the repo, or passes when using both the code from the repo and the new code:

forge::cli config::can_prioritise_closer_lib_remappings (fails with original code and new code) 
forge::cli create::can_create_using_unlocked (fails with original code)
forge::cli config::can_detect_lib_foundry_toml (fails with original code)
forge::cli config::can_override_config (fails with original code)
forge::cli create::can_create_with_constructor_args (fails with original code)
forge::cli cmd::can_init_vscode (fails with original code)
anvil::it traces::test_trace_address_fork (passes with original code and new code) 
anvil::it fork::test_fork_call (passes with original code and new code) 
anvil::it fork::test_fork_timestamp (passes with original code and new code) 
anvil::it otterscan::can_call_ots_get_internal_operations_contract_selfdestruct (passes with original code and new code) 
anvil::it traces::test_trace_address_fork2 (passes with original code and new code) 
forge::cli create::can_create_template_contract (fails with original code)
chisel::cache test_write_session (passes with original code and new code)

**Comment by klkvr** [comment]
yeah, the ci is a little bit broken right now after forgetting-std release, we will fix this soon

**Comment by mattsse** [comment]
is fixed on master

**Comment by matthewliu10** [comment]
I just synced it again

**Comment by matthewliu10** [comment]
I just ran the following tests locally before and after syncing with master. They all passed with both versions.

forge::cli script::can_broadcast_script_skipping_simulation
forge::cli test_cmd::can_disable_block_gas_limit
anvil::it traces::test_trace_address_fork
anvil::it fork::test_fork_timestamp

**Comment by matthewliu10** [comment]
Please review when you get the chance

**Comment by zerosnacks** [comment]
Hi @matthewliu10 

Thanks for your PR! I've pushed some changes to fix the merge conflict (`glob.rs` was moved) and some additional tests

File: crates/config/src/filter.rs [hunk 2]
```diff
@@ -56,6 +56,17 @@ impl GlobMatcher {
             return self.matcher.is_match(format!("./{}", path.display()));
         }
 
+        let glob_path = PathBuf::from(self.glob().glob());
```
Comment (mattsse) [inline crates/config/src/filter.rs:59]: this can be delayed an moved to the && part

we can also use `Path::new` here

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/config/src/filter.rs (hunk 2 above)
Comment (zerosnacks) [reply crates/config/src/filter.rs]: Good catch, applied the suggested changes

**Review by mattsse — APPROVED** [review]

File: crates/forge/tests/cli/test_cmd.rs [hunk 3]
```diff
@@ -180,6 +180,43 @@ contract FailTest is DSTest {
...
+}
+   "#,
+    )
+    .unwrap();
+
+    let test_path = prj.root().join("src/ATest.t.sol");
+    let test_path = test_path.to_string_lossy();
+
+    cmd.args(["test", "--match-path", test_path.as_ref()]);
+    assert!(cmd.stdout_lossy().contains("[PASS]") && !cmd.stdout_lossy().contains("[FAIL]"));
```
Comment (DaniPopes) [inline crates/forge/tests/cli/test_cmd.rs:217]: we can use snapbox here

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/forge/tests/cli/test_cmd.rs (hunk 3 above)
Comment (zerosnacks) [reply crates/forge/tests/cli/test_cmd.rs]: Done, the filter syntax of snapbox is neat :)

**Review by DaniPopes — APPROVED** [review]

**Comment by DaniPopes** [comment]
pending @klkvr

**Review by klkvr — APPROVED** [review]

**Pull request merged**